api/envoy/api/v2/core/protocol.proto

```diff
@@ -42,6 +42,24 @@ message Http1ProtocolOptions {
   // Envoy does not otherwise support HTTP/1.0 without a Host header.
   // This is a no-op if *accept_http_10* is not true.
   string default_host_for_http_10 = 3;
+
+  message HeaderKeyFormat {
+    enum HeaderFormat {
+      // By default, all headers are lower cased.
+      DEFAULT = 0;
+
+      // Formats the header using Train-Case: the first character and any character following a
+      // special character will be capitalized if it's an alpha character. For example,
+      // "content-type" becomes "Content-Type", and "foo$b#$ar" becomes "Foo$B#$Ar".
+      TRAIN_CASE = 1;
+    }
+
+    HeaderFormat header_format = 1;
+  }
+
+  // Describes how the keys for response headers should be formatted. By default, all header keys
+  // are lower cased.
+  HeaderKeyFormat response_header_key_format = 4;
```

Comment:
Is this only for response? What about request headers forwarded to the upstream?

Reply:
i can add it for the request path too, but that seems to be a bit more involved since the http1 codec settings aren't currently being passed through to the connection pool/upstream http/1.1 codec (since all the current options apply only to the HCM). 

How about i add the proto def /w not-implemented and leave that as a TODO?

Reply:
I think it makes sense to have it just be header_format, and keep it hidden until we have it implemented on both sides, because being allowd to set response header format where the settings are used for requests is weird.
WDYT?

Reply:
yeah i like that, will go with that

Reply:
Arguably though clients and backends might have different case senstivity concerns. E.g. one wants Rfc7230-Casing, the other wants 1337Spe4k-k451nG.

Reply:
Sure, but we have codec options in both places.
AFIK when there's protocol options on an HCM, this will affect how we serialize the response headers and when we set the protocol options in CDS it'll affect how we serialize request headers. Since we're not doing enforcement of request headers at the listener or response headers in the connection pool, I don't know what it would mean to have response headers in the connection pool Http1 options config.

I suggest we just have one name, and note it affects how we serialize the headers out at the codec.

Reply:
Yeah, I think the fact that the configuration for request and response headers wouldn't occur in the same location (HCM vs Cluster) makes having it be a single field make more sense than two fields with only one ever being relevant nicer

Reply:
Good point, a single name makes sense given the multiplicity of codec options on downstream/upstream.